Penumbra/UI/SharedTagManager.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Dalamud.Interface;
+using ImGuiNET;
+using OtterGui;
+using OtterGui.Raii;
+using OtterGui.Widgets;
+using Penumbra.Mods;
+using Penumbra.UI.Classes;
+
+namespace Penumbra.UI;
+public sealed class SharedTagManager
+{
+    private static uint _tagButtonAddColor = ColorId.SharedTagAdd.Value();
+    private static uint _tagButtonRemoveColor = ColorId.SharedTagRemove.Value();
+
+    private static float _minTagButtonWidth = 15;
+
+    private const string PopupContext = "SharedTagsPopup";
+    private bool _isPopupOpen = false;
+
+
+    public IReadOnlyList<string> SharedTags { get; internal set; } = Array.Empty<string>();
+
+    public SharedTagManager()
+    {
+    }
+
+    public void ChangeSharedTag(int tagIdx, string tag)
+    {
+        if (tagIdx < 0 || tagIdx > SharedTags.Count)
+            return;
+
+        if (tagIdx == SharedTags.Count) // Adding a new tag
+        {
+            SharedTags = SharedTags.Append(tag).Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
+        }
+        else // Editing an existing tag
+        {
+            var tmpTags = SharedTags.ToArray();
+            tmpTags[tagIdx] = tag;
+            SharedTags = tmpTags.Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
+        }
+    }
+
+    public string DrawAddFromSharedTags(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, bool editLocal)
+    {
+        var tagToAdd = "";
```

Comment:
It probably doesn't cause any differences, but I use `string.Empty` instead of "" wherever possible (i.e. everywhere but default arguments or switches :D )